src/platform/nrfconnect/Reboot.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "Reboot.h"
+
+#include <lib/support/TypeTraits.h>
+
+#include <hal/nrf_power.h>
+#include <zephyr/sys/reboot.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+constexpr uint8_t EncodeReason(SoftwareRebootReason reason)
+{
+    constexpr uint8_t kCustomReasonFlag = 0x80;
+
+    return to_underlying(reason) | kCustomReasonFlag;
+}
+
+void Reboot(SoftwareRebootReason reason)
+{
+    const uint8_t retainedReason = EncodeReason(reason);
```

Comment:
Are retention registers always 8bit length? If so, maybe we can typedef uint8_t to express that somehow. But it's just a nit.

Reply:
I think length of the retention registers is platform-dependent (and is most likely 32-bit), but to common nrfxlib API uses 8 bits only.